### ISSUE TITLE
Inventory update issue

### ISSUE THREAD

**Issue by g-bougard** [glpi-project/glpi#9316]
**Describe the bug**

Asset update with a json inventory doesn't update some components. This also generates some errors in log and an error in agent log as returned response contains php warnings.

**To reproduce**

1. Use glpi-agent to import a computer
2. Use glpi-agent on the same computer to update its inventory
3. Check errors in log

**Expected behavior**

No error expected.
Also in the case a component has changed, I expect the component to be updated but this is not the case.
After investigations, I found the first error in the log came from memory components update. The error happens even if nothing has changed.

**Logs**

```
[2021-07-22 15:33:24] glpiphplog.WARNING: CommonDBTM::getFromDBByCrit() in /var/www/glpi/inc/commondbtm.class.php line 363
getFromDBByCrit expects to get one result, 2 found!  {"user":"inventory@20a321d1fb86"} 
[2021-07-22 15:33:24] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "id" in /var/www/glpi/inc/inventory/asset/device.class.php at line 120
  Backtrace :
  inc/inventory/asset/mainasset.class.php:693        Glpi\Inventory\Asset\Device->handle()
  inc/inventory/asset/mainasset.class.php:630        Glpi\Inventory\Asset\MainAsset->handleAssets()
  inc/ruleimportasset.class.php:910                  Glpi\Inventory\Asset\MainAsset->rulepassed()
  inc/rule.class.php:1437                            RuleImportAsset->executeActions()
  inc/rulecollection.class.php:1469                  Rule->process()
  inc/inventory/asset/mainasset.class.php:489        RuleCollection->processAllRules()
  inc/inventory/inventory.class.php:599              Glpi\Inventory\Asset\MainAsset->handle()
  inc/inventory/inventory.class.php:302              Glpi\Inventory\Inventory->handleItem()
  inc/inventory/request.class.php:139                Glpi\Inventory\Inventory->doInventory()
  inc/inventory/request.class.php:79                 Glpi\Inventory\Request->inventory()
  ...ent/communication/abstractrequest.class.php:257 Glpi\Inventory\Request->handleAction()
  ...ent/communication/abstractrequest.class.php:196 Glpi\Agent\Communication\AbstractRequest->handleJSONRequest()
  front/inventory.php:54                             Glpi\Agent\Communication\AbstractRequest->handleRequest()
  
[2021-07-22 15:33:24] glpiphplog.WARNING: CommonDBTM::getFromDBByCrit() in /var/www/glpi/inc/commondbtm.class.php line 363
getFromDBByCrit expects to get one result, 2 found!  {"user":"inventory@20a321d1fb86","mem_usage":"0.001\", 52.34Mio)"} 
[2021-07-22 15:33:24] glpiphplog.WARNING:   *** PHP Warning (2): Undefined array key "id" in /var/www/glpi/inc/inventory/asset/device.class.php at line 120
  Backtrace :
  inc/inventory/asset/mainasset.class.php:693        Glpi\Inventory\Asset\Device->handle()
  inc/inventory/asset/mainasset.class.php:630        Glpi\Inventory\Asset\MainAsset->handleAssets()
  inc/ruleimportasset.class.php:910                  Glpi\Inventory\Asset\MainAsset->rulepassed()
  inc/rule.class.php:1437                            RuleImportAsset->executeActions()
  inc/rulecollection.class.php:1469                  Rule->process()
  inc/inventory/asset/mainasset.class.php:489        RuleCollection->processAllRules()
  inc/inventory/inventory.class.php:599              Glpi\Inventory\Asset\MainAsset->handle()
  inc/inventory/inventory.class.php:302              Glpi\Inventory\Inventory->handleItem()
  inc/inventory/request.class.php:139                Glpi\Inventory\Inventory->doInventory()
  inc/inventory/request.class.php:79                 Glpi\Inventory\Request->inventory()
  ...ent/communication/abstractrequest.class.php:257 Glpi\Inventory\Request->handleAction()
  ...ent/communication/abstractrequest.class.php:196 Glpi\Agent\Communication\AbstractRequest->handleJSONRequest()
  front/inventory.php:54                             Glpi\Agent\Communication\AbstractRequest->handleRequest()
  
[2021-07-22 15:33:26] glpiphplog.WARNING: CommonDBTM::getFromDBByCrit() in /var/www/glpi/inc/commondbtm.class.php line 363
getFromDBByCrit expects to get one result, 7 found!  {"user":"inventory@20a321d1fb86","mem_usage":"1.571\", 53.18Mio)"} 
[2021-07-22 15:33:26] glpiphplog.WARNING: CommonDBTM::getFromDBByCrit() in /var/www/glpi/inc/commondbtm.class.php line 363
getFromDBByCrit expects to get one result, 3 found!  {"user":"inventory@20a321d1fb86","mem_usage":"0.001\", 53.18Mio)"} 
...
[2021-07-22 15:33:26] glpiphplog.WARNING: CommonDBTM::getFromDBByCrit() in /var/www/glpi/inc/commondbtm.class.php line 363
getFromDBByCrit expects to get one result, 4 found!  {"user":"inventory@20a321d1fb86","mem_usage":"0.002\", 53.18Mio)"} 
[2021-07-22 15:33:26] glpiphplog.WARNING: CommonDBTM::getFromDBByCrit() in /var/www/glpi/inc/commondbtm.class.php line 363
getFromDBByCrit expects to get one result, 4 found!  {"user":"inventory@20a321d1fb86","mem_usage":"0.001\", 53.18Mio)"} 
[2021-07-22 15:33:26] glpiphplog.WARNING:   *** PHP Warning (2): Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in /var/www/glpi/front/inventory.php at line 67
  Backtrace :
  front/inventory.php:67                             header()
  
[2021-07-22 15:33:26] glpiphplog.WARNING:   *** PHP Warning (2): Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in /var/www/glpi/front/inventory.php at line 67
  Backtrace :
  front/inventory.php:67                             header()
  
[2021-07-22 15:33:26] glpiphplog.WARNING:   *** PHP Warning (2): Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in /var/www/glpi/front/inventory.php at line 67
  Backtrace :
  front/inventory.php:67                             header()
  
[2021-07-22 15:33:26] glpiphplog.WARNING:   *** PHP Warning (2): Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in /var/www/glpi/front/inventory.php at line 67
  Backtrace :
  front/inventory.php:67                             header()
  
[2021-07-22 15:33:26] glpiphplog.WARNING:   *** PHP Warning (2): Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in /var/www/glpi/front/inventory.php at line 67
  Backtrace :
  front/inventory.php:67                             header()
```

**glpi-injector output**

```
[guillaume@teclib glpi-agent]$ bin/glpi-injector -u http://localhost:8080/front/inventory.php -f /tmp/3f3e7c30-b296-ffff-b0cc-dd3e10c45be9.json --debug -C
DEBUG: <br />
<b>Warning</b>:  Undefined array key "id" in <b>/var/www/glpi/inc/inventory/asset/device.class.php</b> on line <b>120</b><br />
<br />
<b>Warning</b>:  Undefined array key "id" in <b>/var/www/glpi/inc/inventory/asset/device.class.php</b> on line <b>120</b><br />
<br />
<b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in <b>/var/www/glpi/front/inventory.php</b> on line <b>67</b><br />
<br />
<b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in <b>/var/www/glpi/front/inventory.php</b> on line <b>67</b><br />
<br />
<b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in <b>/var/www/glpi/front/inventory.php</b> on line <b>67</b><br />
<br />
<b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in <b>/var/www/glpi/front/inventory.php</b> on line <b>67</b><br />
<br />
<b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /var/www/glpi/inc/inventory/asset/device.class.php:120) in <b>/var/www/glpi/front/inventory.php</b> on line <b>67</b><br />
{"expiration":24,"status":"ok"}
```

**Your GLPI setup (you can find it in Setup > General menu, System tab)**

```
GLPI 10.0.0-dev ( => /var/www/glpi)
Installation mode: DOCKER
--
Operating system: Linux 20a321d1fb86 5.12.15-200.fc33.x86_64 #1 SMP Wed Jul 7 19:56:49 UTC 2021 x86_64
PHP 8.0.8 apache2handler (Core, PDO, Phar, Reflection, SPL, SimpleXML, apache2handler, apcu, bz2, ctype, curl, date, dom, exif,
 	fileinfo, filter, ftp, gd, hash, iconv, intl, json, ldap, libxml, mbstring, mysqli, mysqlnd, openssl, pcre, pdo_sqlite, posix,
 	session, soap, sodium, sqlite3, standard, tokenizer, xml, xmlreader, xmlwriter, zip, zlib)
Setup: max_execution_time="30" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
 	upload_max_filesize="2M"
Software: Apache/2.4.38 (Debian) (Apache/2.4.38 (Debian) Server at localhost Port 8080)
 	Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:90.0) Gecko/20100101 Firefox/90.0
Server Software: mariadb.org binary distribution
 	Server Version: 10.5.11-MariaDB-1:10.5.11+maria~focal
 	Server SQL Mode: STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
 	Parameters: glpi@db/glpi
 	Host info: db via TCP/IP
...
```

**Additional context**

The following section is what I have in `content/memories` in my json inventory:
```
      "memories": [
         {
            "capacity": 16384,
            "caption": "DIMM_A1",
            "description": "DIMM",
            "manufacturer": "CRUCIAL",
            "model": "BL16G32C16U4B.M16FE",
            "numslots": 1,
            "serialnumber": "E44CBA70",
            "speed": "2666",
            "type": "DDR4"
         },
         {
            "caption": "DIMM_A2",
            "numslots": 2
         },
         {
            "capacity": 16384,
            "caption": "DIMM_B1",
            "description": "DIMM",
            "manufacturer": "CRUCIAL",
            "model": "BL16G32C16U4B.M16FE",
            "numslots": 3,
            "serialnumber": "E44CBA42",
            "speed": "2666",
            "type": "DDR4"
         },
         {
            "caption": "DIMM_B2",
            "numslots": 4
         }
      ],
```

After first import, I have this content in DB for memory devices:
```
MariaDB [glpi]> select * from glpi_devicememories;
+----+-------------+-----------+---------+------------------+--------------+----------------------+-------------+--------------+-----------------------+---------------------+---------------------+
| id | designation | frequence | comment | manufacturers_id | size_default | devicememorytypes_id | entities_id | is_recursive | devicememorymodels_id | date_mod            | date_creation       |
+----+-------------+-----------+---------+------------------+--------------+----------------------+-------------+--------------+-----------------------+---------------------+---------------------+
|  1 | DDR4 - DIMM | 2666      | NULL    |                0 |            0 |                    7 |           0 |            0 |                  NULL | 2021-07-21 15:36:10 | 2021-07-21 15:36:10 |
+----+-------------+-----------+---------+------------------+--------------+----------------------+-------------+--------------+-----------------------+---------------------+---------------------+
1 rows in set (0.000 sec)

MariaDB [glpi]> select * from glpi_items_devicememories;
+----+----------+----------+-------------------+-------+----------+------------+------------+-------------+--------------+-------+-------------+--------------+-----------+
| id | items_id | itemtype | devicememories_id | size  | serial   | is_deleted | is_dynamic | entities_id | is_recursive | busID | otherserial | locations_id | states_id |
+----+----------+----------+-------------------+-------+----------+------------+------------+-------------+--------------+-------+-------------+--------------+-----------+
| 10 |        5 | Computer |                 1 | 16384 | E44CBA70 |          0 |          1 |           0 |            0 | 1     | NULL        |            0 |         0 |
| 11 |        5 | Computer |                 1 | 16384 | E44CBA42 |          0 |          1 |           0 |            0 | 3     | NULL        |            0 |         0 |
+----+----------+----------+-------------------+-------+----------+------------+------------+-------------+--------------+-------+-------------+--------------+-----------+
2 rows in set (0.000 sec)
```

During my analysis, if I comment out lines 119 & 120 in `inc/inventory/asset/device.class.php` which came with 3c3c10db084cf5e5a6567ae42b1ead288151af3d, the error disappears but then if for example I change a memory serial, the serial is not updated in db. Indeed the code on these lines seems to expect only one item device in `glpi_items_devicememories` but we have 2 with the `devicememories_id` set to 1. I'm sure this happens for many other device types as I have a lot of "**getFromDBByCrit expects to get one result**" warnings due to these 2 lines.

**Comment by g-bougard** [comment]
Closes by #9340

**Comment by trasher** [comment]
Should be closed from PR automatically
